### PR TITLE
Feature/la07/add_security

### PR DESCRIPTION
**Pull request by AliaksandrAnashkevich**

### CODE REVIEW

File: src/main/java/com/academia/library/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.cors().and()
+                .csrf().disable()
+                .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+                .authorizeRequests()
+                .antMatchers("/").permitAll()
+                .antMatchers(HttpMethod.POST, "/login").permitAll()
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/config/SecurityConfig.java:41]: Please move to constants

File: src/main/java/com/academia/library/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.cors();
+        http.csrf().disable();
+        http.sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);
+        http.authorizeRequests()
+                .antMatchers("/swagger-ui/**").permitAll()
+                .antMatchers("/login").permitAll()
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/config/SecurityConfig.java]: Please move to constants

File: src/main/java/com/academia/library/controller/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+@RestController
+@RequestMapping("/")
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/controller/AuthController.java]: Why do you use it here ?

File: src/main/java/com/academia/library/cryptor/impl/UserCryptorImpl.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@Component
+public class UserCryptorImpl implements UserCryptor {
+
+    @Override
+    public String decode(String encodeString) {
+        return new String(Base64.decodeBase64(encodeString.getBytes(StandardCharsets.UTF_8)));
+    }
+
+    @Override
+    public String encode(String string) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/cryptor/impl/UserCryptorImpl.java]: What does this name mean ?

File: src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.academia.library.exception;
+
+public class InvalidAuthRequestDataException extends RuntimeException {
+
+    private static final String MESSAGE_CONSTRUCTION = "Invalid email or password.";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java]: What does this name mean ?

File: src/main/java/com/academia/library/model/User.java [hunk 6]
```diff
@@ -45,9 +49,11 @@
     @Column
     private String password;
 
+    @Generated(GenerationTime.INSERT)
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/model/User.java]: Please use liqubase for insert date

File: src/main/java/com/academia/library/security/JwtTokenFilter.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@RequiredArgsConstructor
+public class JwtTokenFilter extends OncePerRequestFilter {
+
+    private final JwtTokenProvider jwtTokenProvider;
+    private final AuthenticationEntryPoint authenticationEntryPoint;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+                                    FilterChain filterChain) throws ServletException, IOException {
+        try {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/security/JwtTokenFilter.java:27]: Please use formating

File: src/main/java/com/academia/library/security/JwtTokenProvider.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
+package com.academia.library.security;
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/security/JwtTokenProvider.java:2]: Please delete empty line

File: src/main/resources/liquibase/changeset/20.01.2022_1505_insert_users_roles_authority_data.xml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            BEGIN TRANSACTION;
+
+            INSERT INTO user_roles(user_id, role_id) VALUES
+            ((SELECT id FROM users WHERE email = 'admin@gmail.com'), (SELECT id FROM roles WHERE name = 'ROLE_ADMIN'));
+
+            INSERT INTO user_roles(user_id, role_id) VALUES
+            ((SELECT id FROM users WHERE email = 'user@gmail.com'), (SELECT id FROM roles WHERE name = 'ROLE_USER'));
+
+            COMMIT;
+        </sql>
```
Comment (DenisKiparizz) [inline src/main/resources/liquibase/changeset/20.01.2022_1505_insert_users_roles_authority_data.xml]: Please use insert like bellow

File: src/main/java/com/academia/library/config/SecurityConfig.java (hunk 2 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/config/SecurityConfig.java]: Ok, thank

File: src/main/java/com/academia/library/controller/AuthController.java (hunk 3 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/controller/AuthController.java]: Ok, i fix this moment

File: src/main/java/com/academia/library/cryptor/impl/UserCryptorImpl.java (hunk 4 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/cryptor/impl/UserCryptorImpl.java]: Ok, thank

File: src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java (hunk 5 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java]: Ok, thank

File: src/main/java/com/academia/library/model/User.java (hunk 6 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/model/User.java]: Ok, thank

File: src/main/java/com/academia/library/security/JwtTokenFilter.java (hunk 7 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/security/JwtTokenFilter.java:27]: Ok, thank

File: src/main/java/com/academia/library/security/JwtTokenProvider.java (hunk 8 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/security/JwtTokenProvider.java:2]: Ok, thank

File: src/main/resources/liquibase/changeset/20.01.2022_1505_insert_users_roles_authority_data.xml (hunk 9 above)
Comment (AliaksandrAnashkevich) [reply src/main/resources/liquibase/changeset/20.01.2022_1505_insert_users_roles_authority_data.xml]: Ok, thank

File: src/main/java/com/academia/library/config/SecurityConfig.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public AuthenticationManager authenticationManagerBean() throws Exception {
+        return super.authenticationManagerBean();
+    }
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder(12);
+    }
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/config/SecurityConfig.java]: Please remove empty line

File: src/main/java/com/academia/library/config/SecurityConfig.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+@RequiredArgsConstructor
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    private final JwtTokenFilter jwtTokenFilter;
+    private final JwtUserDetailsService userDetailsService;
+    private static final String SWAGGER_ENDPOINT = "/swagger-ui/**";
+    private static final String LOGIN_ENDPOINT = "/auth/login";
+    private static final String REGISTRATION_ENDPOINT = "/auth/registration";
+    private static final String BOOK_ENDPOINT = "/books/**";
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/config/SecurityConfig.java]: Please move this constants above
Above this

private final JwtTokenFilter jwtTokenFilter;
private final JwtUserDetailsService userDetailsService;

File: src/main/java/com/academia/library/config/SecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    protected void configure(final AuthenticationManagerBuilder auth) throws Exception {
+        auth
+                .userDetailsService(userDetailsService)
+                .passwordEncoder(passwordEncoder());
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/config/SecurityConfig.java]: Do in one line

File: src/main/java/com/academia/library/exception/handler/AuthExceptionHandler.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity handleValidationExceptions(MethodArgumentNotValidException e) {
+        Map<String, String> errors = new HashMap<>();
+        e.getBindingResult().getAllErrors().forEach((error) -> {
+            String fieldName = ((FieldError) error).getField();
+            String errorMessage = error.getDefaultMessage();
+            errors.put(fieldName, errorMessage);
+        });
+        log.error("Bad Request {}", errors);
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/handler/AuthExceptionHandler.java]: Please refactor this 
It looks doesn't brilliant

File: src/main/java/com/academia/library/exception/BookNotFoundException.java [hunk 14]
```diff
@@ -2,9 +2,9 @@
 
 public class BookNotFoundException extends RuntimeException {
 
-    private static final String MESSAGE_CONSTRUCTOR = "Book with id=[%s] don't exist";
+    private static final String MESSAGE_ERROR = "Book with id=[%s] don't exist";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/BookNotFoundException.java]: Rename to BOOK_NOT_FOUND_MESSAGE

File: src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package com.academia.library.exception;
+
+public class InvalidAuthRequestDataException extends RuntimeException {
+
+    private static final String MESSAGE_ERROR = "Invalid email or password.";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java]: Rename to INVALID_AUTH_MESSAGE

File: src/main/java/com/academia/library/exception/JwtTokenException.java [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package com.academia.library.exception;
+
+public class JwtTokenException extends RuntimeException{
+
+    private static final String MESSAGE_ERROR = "Expired or invalid JWT token";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/JwtTokenException.java]: Rename to INVALID_JWT_MESSAGE

File: src/main/java/com/academia/library/exception/RoleNotFoundException.java [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package com.academia.library.exception;
+
+public class RoleNotFoundException extends RuntimeException {
+
+    private static final String MESSAGE_ERROR = "Role with name=[%s] don't exist";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/RoleNotFoundException.java]: Rename to ROLE_NOT_FOUND_MESSAGE

File: src/main/java/com/academia/library/exception/UserNotFoundException.java [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package com.academia.library.exception;
+
+public class UserNotFoundException extends RuntimeException {
+
+    private static final String MESSAGE_ERROR = "User with username=[%s] don't found";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/UserNotFoundException.java]: Rename to USER_NOT_FOUND_MESSAGE

File: src/main/java/com/academia/library/mapper/UserMapper.java [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.academia.library.dto.UserRequestDto;
+import com.academia.library.dto.UserResponseDto;
+import com.academia.library.model.User;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring")
+public interface UserMapper {
+
+    UserResponseDto toDto(User user);
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/mapper/UserMapper.java]: Please rename UserResponseDto to UserResponse
Do that for ALL similar names

File: src/main/java/com/academia/library/security/JwtAuthenticationEntryPoint.java [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Component
+public class JwtAuthenticationEntryPoint implements AuthenticationEntryPoint {
+    @Override
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/security/JwtAuthenticationEntryPoint.java:14]: Please add empty line here

File: src/main/java/com/academia/library/util/RoleNames.java [hunk 21]
```diff
@@ -0,0 +1,7 @@
+package com.academia.library.util;
+
+public class RoleNames {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/util/RoleNames.java:3]: Please apply formatting here

File: src/main/resources/liquibase/changeset/06.01.2022_1837_create_tags_table.xml [hunk 22]
```diff
@@ -4,7 +4,7 @@
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.1.xsd">
-    <changeSet id="14.01.2022_1416_create_tags_sequence" author="aliaksandr anashkevich">
+    <changeSet id="06.01.2022_1836_create_tags_sequence" author="aliaksandr anashkevich">
```
Comment (DenisKiparizz) [inline src/main/resources/liquibase/changeset/06.01.2022_1837_create_tags_table.xml]: Lets rename date
YYYY.MM.DD_TIME_DESCRIPTION

File: src/test/java/com/academia/library/controller/AuthControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    @AfterEach
+    private void deleteAll() {
+        userRepository.deleteAll();
+        roleRepository.deleteAll();
+        authorityRepository.deleteAll();
+    }
+
+    @DisplayName("Login user")
```
Comment (DenisKiparizz) [inline src/test/java/com/academia/library/controller/AuthControllerTest.java:97]: Please add this one everywhere

File: src/test/java/com/academia/library/LibraryApplicationTests.java [hunk 24]
```diff
@@ -2,7 +2,9 @@
 
 import org.junit.jupiter.api.Test;
 import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ActiveProfiles;
 
+@ActiveProfiles("test")
 @SpringBootTest
 class LibraryApplicationTests {
```
Comment (DenisKiparizz) [inline src/test/java/com/academia/library/LibraryApplicationTests.java]: Do you need this test ?

File: build.gradle [hunk 25]
```diff
@@ -14,32 +14,41 @@ configurations {
     }
 }
 
+
```
Comment (DenisKiparizz) [inline build.gradle]: Please remove empty line

File: src/test/java/com/academia/library/LibraryApplicationTests.java (hunk 24 above)
Comment (AliaksandrAnashkevich) [reply src/test/java/com/academia/library/LibraryApplicationTests.java]: Yes

File: build.gradle (hunk 25 above)
Comment (AliaksandrAnashkevich) [reply build.gradle]: Ok, thank

File: src/test/java/com/academia/library/controller/AuthControllerTest.java (hunk 23 above)
Comment (AliaksandrAnashkevich) [reply src/test/java/com/academia/library/controller/AuthControllerTest.java:97]: Ok, 
I add this moment in Endpoint task

File: src/main/resources/liquibase/changeset/06.01.2022_1837_create_tags_table.xml (hunk 22 above)
Comment (AliaksandrAnashkevich) [reply src/main/resources/liquibase/changeset/06.01.2022_1837_create_tags_table.xml]: Ok, thank

File: src/main/java/com/academia/library/util/RoleNames.java (hunk 21 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/util/RoleNames.java:3]: Ok, thank

File: src/main/java/com/academia/library/config/SecurityConfig.java (hunk 12 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/config/SecurityConfig.java]: Ok, thank

File: src/main/java/com/academia/library/config/SecurityConfig.java (hunk 11 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/config/SecurityConfig.java]: Ok, thank

File: src/main/java/com/academia/library/config/SecurityConfig.java (hunk 10 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/config/SecurityConfig.java]: Ok, thank

File: src/main/java/com/academia/library/exception/handler/AuthExceptionHandler.java (hunk 13 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/handler/AuthExceptionHandler.java]: Ok, thank

File: src/main/java/com/academia/library/exception/BookNotFoundException.java (hunk 14 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/BookNotFoundException.java]: Ok, thank

File: src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java (hunk 15 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/InvalidAuthRequestDataException.java]: Ok, thank

File: src/main/java/com/academia/library/exception/JwtTokenException.java (hunk 16 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/JwtTokenException.java]: Ok, thank

File: src/main/java/com/academia/library/exception/RoleNotFoundException.java (hunk 17 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/RoleNotFoundException.java]: Ok, thank

File: src/main/java/com/academia/library/exception/UserNotFoundException.java (hunk 18 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/exception/UserNotFoundException.java]: Ok, thank

File: src/main/java/com/academia/library/mapper/UserMapper.java (hunk 19 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/mapper/UserMapper.java]: Ok, thank

File: src/main/java/com/academia/library/security/JwtAuthenticationEntryPoint.java (hunk 20 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/security/JwtAuthenticationEntryPoint.java:14]: Ok, thank

File: src/main/java/com/academia/library/controller/AuthController.java [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Operation(summary = "Registration", description = "Registration new user")
+    @ApiResponse(responseCode = "200", description = "Success",
+            content = @Content(mediaType = "application/json", schema = @Schema(implementation = BookResponse.class)))
+    @ApiResponse(responseCode = "400", description = "Invalid user request body", content = @Content)
+    @PostMapping("/registration")
+    public UserResponse registration(@Valid @RequestBody UserRequest userRequest) {
+        return userService.create(userRequest);
+    }
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/controller/AuthController.java]: Please remove empty line

**Review by DenisKiparizz — APPROVED** [review]

File: src/main/java/com/academia/library/controller/AuthController.java (hunk 26 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/controller/AuthController.java]: Ok, thanks

**Pull request merged**